web/preferences.js

```diff
@@ -54,27 +44,21 @@ class BasePreferences {
 
     this.#initializedPromise = this._readFromStorage(this.#defaults).then(
       ({ browserPrefs, prefs }) => {
-        const options = Object.create(null);
-
-        for (const [name, val] of Object.entries(this.#browserDefaults)) {
-          const prefVal = browserPrefs?.[name];
-          options[name] = typeof prefVal === typeof val ? prefVal : val;
-        }
-        for (const [name, val] of Object.entries(this.#defaults)) {
-          const prefVal = prefs?.[name];
-          // Ignore preferences whose types don't match the default values.
-          options[name] = this.#prefs[name] =
-            typeof prefVal === typeof val ? prefVal : val;
+        if (
+          (typeof PDFJSDev === "undefined" || PDFJSDev.test("GENERIC")) &&
+          AppOptions._checkDisablePreferences()
+        ) {
+          return;
         }
-        AppOptions.setAll(options, /* init = */ true);
+        AppOptions.setAll({ ...browserPrefs, ...prefs }, /* prefs = */ true);
       }
     );
 
     if (typeof PDFJSDev !== "undefined" && PDFJSDev.test("MOZCENTRAL")) {
-      window.addEventListener("updatedPreference", evt => {
-        this.#updatePref(evt.detail);
+      window.addEventListener("updatedPreference", async evt => {
+        await this.#initializedPromise;
+        AppOptions.setAll(evt.detail, /* prefs = */ true);
```

Comment:
Looking at the current code `evt.detail` here is an object like `{ name: 'foo', value: 'bar' }` looking at the `#updatePref` method's parameter, because it's a single preference being updated, but `AppOptions.setAll` seems to expect an object like `{ 'foo': 'bar' }` unless I'm mistaken, so does this work as expected?

Reply:
Looking at this again I'm not sure, so I'll test this in a Firefox build to check if it's correct.

Reply:
Correct, this didn't work as intended; really good catch!

---

I suppose this is a good example or why it's never a good idea to attempt to simplify code *after* you thing that you're done with testing :-)